noxfile.py

```diff
@@ -17,20 +24,193 @@
     "clang-format",
 ]
 
+#
+# Helpers for use within this file.
+#
+
+
+def simulator_support_matrix() -> List[Tuple[str, str, str]]:
+    """
+    Get a list of supported simulator/toplevel-language/GPI-interface tuples.
+    """
+
+    # Simulators with support for VHDL through VHPI, and Verilog through VPI.
+    standard = [
+        (sim, toplevel_lang, gpi_interface)
+        for sim in ("activehdl", "rivierapro", "xcelium")
+        for toplevel_lang in ("verilog", "vhdl")
+        for gpi_interface in ("vpi", "vhpi")
+        if (toplevel_lang, gpi_interface) in (("verilog", "vpi"), ("vhdl", "vhpi"))
+    ]
+
+    # Special-case simulators.
+    special = [
+        ("cvc", "verilog", "vpi"),
+        ("ghdl", "vhdl", "vpi"),
+        ("icarus", "verilog", "vpi"),
+        ("questa", "verilog", "vpi"),
+        ("questa", "vhdl", "fli"),
+        ("questa", "vhdl", "vhpi"),
+        ("verilator", "verilog", "vpi"),
+        ("vcs", "verilog", "vpi"),
+    ]
+
+    return standard + special
+
+
+def env_vars_for_test(
+    sim: str, toplevel_lang: str, gpi_interface: str
+) -> Dict[str, str]:
+    """Prepare the environment variables controlling the test run."""
+    e = {}
+    if sim is not None:
+        e["SIM"] = sim
+    if toplevel_lang is not None:
+        e["TOPLEVEL_LANG"] = toplevel_lang
+    assert not (toplevel_lang == "verilog" and gpi_interface != "vpi")
+    if toplevel_lang == "vhdl":
+        e["VHDL_GPI_INTERFACE"] = gpi_interface
+
+    return e
+
+
+def stringify_dict(d: Dict[str, str]) -> str:
+    return ", ".join(f"{k}={v}" for k, v in d.items())
+
+
+#
+# Development pipeline
+#
+# - Use nox to build an sdist; no separate build step is required.
+# - Run tests against the installed sdist.
+# - Collect coverage.
+#
+
+
+@nox.session
+def dev_build(session: nox.Session) -> None:
+    session.warning("No building is necessary for development sessions.")
+
+
+@nox.session
+def dev_test(session: nox.Session) -> None:
+    """Run all development tests as configured through environment variables."""
+
+    # Not using session.notify() here to be able to pass parameters to the
+    # session.
+    dev_test_pytest(session)
+    dev_test_sim(session, sim=None, toplevel_lang=None, gpi_interface=None)
+
 
 @nox.session
-def tests(session: nox.Session) -> None:
-    """run cocotb regression suite"""
+@nox.parametrize("sim,toplevel_lang,gpi_interface", simulator_support_matrix())
```

Comment:
Since we aren't ever running the full matrix in any environment I don't see the value here. The environment variables are more flexible and less work. gpi_interface will change in the future, so this creates future work for ourselves.

Reply:
This is actually a feature I have been waiting for for a while. Three points:
- If gpi_interface changes, this code has to change as well. That's fine, it's just code. (And gpi_interface doesn't change regularly.)
- Having a test depend on the environment makes it harder to reproduce, and easier to test the wrong thing. Good tests are isolated from that; that's what tox and nox are trying hard to do, and something we break by relying on those environment variables for test selection.
- I regularly get annoyed if jobs which are run in CI are hard to reproduce locally. There should simply be no magic in a CI setup, it should all be in scripting that I can run locally as well. Take two examples:
  - When preparing for a release I want to e.g. test against three simulators that I have installed locally, and let that just run overnight. With the new code I can do that with a single command line. With the existing setup, I'd have to write a wrapper bash script that sets the environment variables and then kicks of a run.
  - I want to give a single command line to an EDA vendor for them to test cocotb exhaustively against their simulator, with all interfaces they support. We can now do that.

Reply:
`parameterize` is intended for running the cross product of the values for each parameter in testing, at least that's what it's used for in pytest. The ability to run only a limited cross section is not the intended way to test in pytest. So either nox is misapplying the feature, or we are. In either case it seems out needs have already expanded past the capabilities of nox unless we resort to abusing features. 

I'm willing to ignore this to get the wheel build stuff in quickly.

Reply:
Leaving the intention-guessing aside: 
* `parametrize` is public API of nox, and documented to be used in a similar way as we use it (have a look at https://nox.thea.codes/en/stable/config.html#parametrizing-the-session-python). 
* Using `parametrize` is beneficial over using argparse because we 
  * have all valid parameter combinations showing up in the help output (`nox -l`), and 
  * can use `nox -k` to select groups of combinations easily.
  We could of course implement similar or identical functionality ourselves with argparse, at the cost of duplicating functionality that already exists in nox.

Reply:
I'm not sure it's intention guessing, the default behavior runs everything, which clearly indicates the intended use case. But if `nox` thinks that's a valid way to use it, and `tox-gh-actions` did the same, then I guess it's general understanding though I'm not a fan.

Perhaps `nox` needs to implement something like click's `option` decorator. There is a clear distinction between the use cases of tasks for project automation vs CI testing. It seems to be trying to both at the same time, and not cleanly.

Reply:
Maybe that's the root of our disagreement. You wrote:
> the default behavior runs everything

I would argue running tests against every supported simulator is a good default, also for our use case. A full (acceptance) test run of cocotb should run against all supported simulators. If we run against a subset of simulators, we also run a subset of a full test run.